### PR TITLE
[jax2tf] General conversion of reduce_window.

### PR DESCRIPTION
**Pull request by bchetioui**
Much like scatter_p, the conversion works as well as the underlying
reduction function (e.g. reduce_window_max is not properly converted
for the int8 dtype, because tf.math.maximum is not defined for int8
tensors).

### CODE REVIEW

File: jax/experimental/jax2tf/jax2tf.py [hunk 1]
```diff
@@ -882,11 +883,14 @@ def _get_shape_from_tensor_or_array(x):
...
 
 
-def _reduce_window(jax_f, reducer, init_val, operand, window_dimensions,
-                   window_strides, padding, base_dilation, window_dilation,
-                   input_shape=None):
+def _specialized_reduce_window(jax_f, reducer, identity, operand,
+                               window_dimensions, window_strides, padding,
+                               base_dilation, window_dilation,
+                               input_shape=None):
   """TensorFlow implementation of reduce_window_{sum,min,max}."""
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:890]: Please write a docstring documenting the usage

File: jax/experimental/jax2tf/jax2tf.py [hunk 2]
```diff
@@ -904,13 +908,66 @@ def _reduce_window(jax_f, reducer, init_val, operand, window_dimensions,
...
+def _get_max_identity(dtype):
+  jax_dtype = to_jax_dtype(dtype)
+  val = None
+  if dtypes.issubdtype(jax_dtype, np.inexact):
+    val = -np.inf
+  elif dtypes.issubdtype(jax_dtype, np.integer):
+    val = dtypes.iinfo(jax_dtype).min
+  elif dtypes.issubdtype(jax_dtype, np.bool_):
+    val = False
+  return _safe_convert_to_tensor(val, dtype)
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:921]: Maybe change the code to make it obvious that val cannot be None.

File: jax/experimental/jax2tf/jax2tf.py [hunk 3]
```diff
@@ -904,13 +908,66 @@ def _reduce_window(jax_f, reducer, init_val, operand, window_dimensions,
...
+def _get_min_identity(dtype):
+  jax_dtype = to_jax_dtype(dtype)
+  val = None
+  if dtypes.issubdtype(jax_dtype, np.inexact):
+    val = np.inf
+  elif dtypes.issubdtype(jax_dtype, np.integer):
+    val = dtypes.iinfo(jax_dtype).max
+  elif dtypes.issubdtype(jax_dtype, np.bool_):
+    val = True
+  return _safe_convert_to_tensor(val, dtype)
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:932]: As above, val cannot be None.

File: jax/experimental/jax2tf/jax2tf.py [hunk 4]
```diff
@@ -904,13 +908,66 @@ def _reduce_window(jax_f, reducer, init_val, operand, window_dimensions,
                             window_dilations=window_dilation, padding=padding)
   out.set_shape(out_shape)
   return out
+
+def _get_max_identity(dtype):
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:912]: Is the return type a `TfVal` or a jax value?

Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:912]: Let's rename this to `tf_dtype`. Perhaps the whole function can be written using only TF dtypes?

File: jax/experimental/jax2tf/jax2tf.py [hunk 5]
```diff
@@ -904,13 +908,66 @@ def _reduce_window(jax_f, reducer, init_val, operand, window_dimensions,
...
+  out_shape = _reduce_window_shape(lax._reduce_window_sum, operand, window_dimensions,
+                                   window_strides, padding, base_dilation,
+                                   window_dilation)
+
+  def reducer(arg1: TfVal, arg2: TfVal) -> TfVal:
+    typed_jaxpr = _mk_typed_jaxpr(jaxpr, consts)
+    res, = _interpret_jaxpr(typed_jaxpr, arg1, arg2)
+    return res
+
+  o_spec = tf.TensorSpec(operand.shape, dtype=operand.dtype)
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:948]: It seems that everything below here can be shared with _Specialized_reduce_window?

File: jax/experimental/jax2tf/tests/primitives_test.py [hunk 6]
```diff
@@ -279,20 +279,37 @@ def test_select_and_gather_add(self, harness: primitive_harness.Harness):
...
     dtype = harness.params['dtype']
 
-    safe_computations = [('sum', dtype(0))]
+    expect_tf_exceptions = False
 
-    if dtype in jtu.dtypes.all_floating:
-      # Only in this case, np.inf can be casted safely to a meaningful value.
-      safe_computations += [('max', dtype(-np.inf)), ('min', dtype(np.inf))]
+    if (jtu.device_under_test() == 'tpu' and dtype is np.complex64):
+      raise unittest.SkipTest('TODO: reduce_window on TPU does not handle complex64')
```
Comment (gnecula) [inline jax/experimental/jax2tf/tests/primitives_test.py:288]: Please specify that this is true in JAX

File: jax/experimental/jax2tf/jax2tf.py (hunk 2 above)
Comment (SIben) [reply jax/experimental/jax2tf/jax2tf.py]: Addressed in `eee7daa`

File: jax/experimental/jax2tf/jax2tf.py (hunk 3 above)
Comment (SIben) [reply jax/experimental/jax2tf/jax2tf.py]: Addressed in `eee7daa`

File: jax/experimental/jax2tf/jax2tf.py (hunk 4 above)
Comment (SIben) [reply jax/experimental/jax2tf/jax2tf.py]: Addressed in `eee7daa`

Comment (SIben) [reply jax/experimental/jax2tf/jax2tf.py]: `TfVal`: Addressed in `eee7daa`

File: jax/experimental/jax2tf/tests/primitives_test.py (hunk 6 above)
Comment (SIben) [reply jax/experimental/jax2tf/tests/primitives_test.py]: Addressed in `6c8554e7`

File: jax/experimental/jax2tf/jax2tf.py (hunk 5 above)
Comment (SIben) [reply jax/experimental/jax2tf/jax2tf.py]: Addressed in `54ac20e4`

File: jax/experimental/jax2tf/jax2tf.py (hunk 1 above)
Comment (SIben) [reply jax/experimental/jax2tf/jax2tf.py]: Addressed in `3b8ae7ab`

File: jax/experimental/jax2tf/jax2tf.py [hunk 7]
```diff
@@ -904,13 +899,97 @@ def _reduce_window(jax_f, reducer, init_val, operand, window_dimensions,
...
+
+def _specialized_reduce_window(reducer, identity, operand, window_dimensions,
+                               window_strides, padding, base_dilation,
+                               window_dilation):
+  """Wraps the TensorFlow reduce window operation based on a reducer and an identity
+  function defining the initial value of the reduction depending on the dtype of the
+  operand.
+
+  Args:
+    reducer: reduction function of type T -> T -> T
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:941]: I would change T with TfVal. It is important that this is a TF function.

File: jax/experimental/jax2tf/jax2tf.py [hunk 8]
```diff
@@ -904,13 +899,97 @@ def _reduce_window(jax_f, reducer, init_val, operand, window_dimensions,
...
+    identity: function that takes a dtype as a parameter and returns the starting
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:942]: Here `dtype` is `tf_dtype`

File: jax/experimental/jax2tf/jax2tf.py [hunk 9]
```diff
@@ -904,13 +899,97 @@ def _reduce_window(jax_f, reducer, init_val, operand, window_dimensions,
...
+    The reduced operand.
+  """
+
+  return _common_reduce_window(
+      operand, identity(operand.dtype), reducer, window_dimensions,
+      window_strides, padding, base_dilation, window_dilation
+  )
+
+def _get_max_identity(tf_dtype):
+  numpy_tf_dtype = tf_dtype.as_numpy_dtype
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:961]: This is better now, it is clear that it deals only with TF stuff

**Review by gnecula — APPROVED** [review]

File: jax/experimental/jax2tf/jax2tf.py (hunk 7 above)
Comment (SIben) [reply jax/experimental/jax2tf/jax2tf.py]: Addressed in `e40fdea6`

File: jax/experimental/jax2tf/jax2tf.py (hunk 8 above)
Comment (SIben) [reply jax/experimental/jax2tf/jax2tf.py]: Addressed in `e40fdea6`

**Pull request merged**